### PR TITLE
 InstallableFlake::toDerivedPaths(): Support paths and store paths

### LINKED ISSUE #7417: allow installables without drvPath
Installables that point to a builtins.storePath or builtins.fetchClosure cannot be used from the command line due to not having a drvPath, yet they can be used if wrapped in a builtins:buildenv or any other derivation wrapper.

**Describe the solution you'd like**
Allow usage of installables (even flake references) without requiring them to have drvPath's.

**Describe alternatives you've considered**
- Wrapping each such installable in a builtins:buildenv, which works.
- I tried following the source. It seems that goals need a name, and for attribute/flakerefs it needs a drvPath to "name" the goal, but otherwise it looks like an artificial limitation.
- `nix build $(nix eval --impure .#some-attr-with-builtins-storePath)` works, which is odd because one would think this could be done automatically

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by edolstra**
This makes `nix build` work on paths (which will be copied to the store) and store paths (returned as is). E.g. the following flake output attributes can be built using `nix build .#foo`:

```
foo = ./src;
foo = self.outPath;
foo = builtins.fetchTarball { ... };
foo = (builtins.fetchTree { .. }).outPath;
foo = builtins.fetchTree { .. } + "/README.md";
foo = builtins.storePath /nix/store/...;
```

Note that this is potentially risky, e.g.

```
foo = /.;
```

will cause Nix to try to copy the entire file system to the store.

What doesn't work yet:

```
foo = self;
foo = builtins.fetchTree { .. };
```

because we don't handle attrsets with an outPath attribute in it yet, and

```
foo = builtins.storePath /nix/store/.../README.md;
```

since result symlinks have to point to a store path currently (rather than a file inside a store path).

Extracted from the lazy-trees branch. Fixes #7417. This PR also refactors `Installables` to merge `InstallableFlake::toDerivations()` into `toDerivedPaths()`, which simplifies the interface.

### CODE REVIEW

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nix/pull/7543 This should help us handle the remaining cases.

**Review by Ericson2314 — APPROVED** [review]
I think this PR is an excellent first step, and we should merge it as is, but we should take away the "fixes #7543" because there is more to that issue remaining.

I am happy to do the rest myself.

**Review by thufschmitt — CHANGES_REQUESTED** [review]
That's a great addition!

**Comment by tomberek** [comment]
Looking to include the more advanced features to support:
- attrsets with outPath
- derivations without drvPath
- derivation without drvPath with multiple outputs with strings
- - derivation without drvPath with multiple outputs with paths
- 
https://github.com/NixOS/nix/compare/master...flox:nix:fix-7417 (WIP)

File: src/nix/profile.cc [hunk 1]
```diff
@@ -476,18 +481,22 @@ struct CmdProfileUpgrade : virtual SourceExprCommand, MixDefaultProfile, MixProf
...
+                assert(info.resolvedRef && info.attrPath);
 
-                if (element.source->resolvedRef == resolvedRef) continue;
+                if (element.source->resolvedRef == info.resolvedRef) continue;
 
                 printInfo("upgrading '%s' from flake '%s' to '%s'",
-                    element.source->attrPath, element.source->resolvedRef, resolvedRef);
+                    element.source->attrPath, element.source->resolvedRef, *info.resolvedRef);
 
                 element.source = ProfileElementSource {
```
[the remark this answers is not in the document]
Comment (edolstra) [reply src/nix/profile.cc:495]: They largely contain the same info now, but that might not be the case in the future.

File: src/nix/profile.cc [hunk 2]
```diff
@@ -251,13 +253,19 @@ struct ProfileManifest
...
 {
-    std::map<Installable *, BuiltPaths> res;
-    for (auto & [installable, builtPath] : builtPaths)
-        res[installable.get()].push_back(builtPath.path);
+    std::map<Installable *, std::pair<BuiltPaths, ExtraInfo>> res;
+    for (auto & [installable, builtPath] : builtPaths) {
+        auto & r = res[installable.get()];
+        /* Note that there could be conflicting info
+           (e.g. meta.priority fields) if the installable returned
+           multiple derivations. So pick one arbitrarily. */
```
[the remark this answers is not in the document]
Comment (edolstra) [reply src/nix/profile.cc]: I've added a FIXME me for this.

File: src/nix/build.cc [hunk 3]
```diff
@@ -94,13 +94,15 @@ struct CmdBuild : InstallablesCommand, MixDryRun, MixJSON, MixProfile
...
             std::vector<DerivedPath> pathsToBuild;
 
-            for (auto & i : installables) {
-                auto b = i->toDerivedPaths();
-                pathsToBuild.insert(pathsToBuild.end(), b.begin(), b.end());
-            }
+            for (auto & i : installables)
+                for (auto & b : i->toDerivedPaths())
+                    pathsToBuild.push_back(b.path);
+
```
[the remark this answers is not in the document]
Comment (edolstra) [reply src/nix/build.cc:100]: IMHO in C++ there isn't much risk here (unlike e.g. in a dynamically scoped language like Python).

**Comment by tomberek** [comment]
> * attrsets with outPath
> * derivations without drvPath
> * derivation without drvPath with multiple outputs with strings
> * derivation without drvPath with multiple outputs with paths

@edolstra would you want the above implemented in a new PR?

File: tests/flakes/build-paths.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    a2 = ./foo;
+
+    a3 = ./.;
+
+    a4 = self.outPath;
+
+    # FIXME
+    a5 = self;
+
+    a6 = flake2.outPath;
```
Comment (Ericson2314) [inline tests/flakes/build-paths.sh:35]: How does this work? This should have an `!out!drvPath` string context.

**Pull request merged**